### PR TITLE
Adding more Spawner Features.

### PR DESCRIPTION
**Pull request by Asleeepp**
### Description

This PR aims to add a ton more functionality for both regular spawners, and the newly added trial spawners. See below for a list of tasks.

- [x] Add Spawner and PreSpawner event.
- [x] Reset spawner time effect
- [x] required player range expression
- [x] max nearby entities expression
- [x] spawn count, spawn range and spawn delay expression
- [x] add tests for all above ^ (jesus christ)

Trial Spawner (other PR)
- [ ] cooldown length expression
- [ ] tracked players & entities expression and condition
- [ ] tracked players & entities stop/start effect
- [ ] ominous expression
- [ ] normal/ominous configuration expression(?)
- [ ] base spawns before cooldown expression
- [ ] base simultaneous entities expression
- [ ] possible rewards expression(?)
- [ ] get trial spawner state expression

Above is my plans for this PR. (its quite a lot i know, please tell me if you have any doubts/objections)

---
**Target Minecraft Versions:** any 
**Requirements:** none 
**Related Issues:** https://github.com/SkriptLang/Skript/issues/6856

### CODE REVIEW

**Comment by sovdeeth** [comment]
Should also cover the potential spawns api (perhaps with full entity snapshot support?). Also highly suggest doing this via multiple prs that all target one branch (i can make a feature/spawners branch if you'd like). It'll make it easier to review and more organized.

**Comment by Asleeepp** [comment]
> Should also cover the potential spawns api (perhaps with full entity snapshot support?). Also highly suggest doing this via multiple prs that all target one branch (i can make a feature/spawners branch if you'd like). It'll make it easier to review and more organized.

yeah, that's nice. i'll do that if you make the branch, just need to figure out what i want to separate, i think i'll keep this PR open with the tasks and update accordingly when i make new PRs, this'll probably just be the base spawner stuff, and the other PRs can be for trial spawners/BaseSpawner

**Comment by sovdeeth** [comment]
there ya go

**Comment by Pikachu920** [comment]
> Should also cover the potential spawns api (perhaps with full entity snapshot support?).

not sure if this should be a requirement for merge. it's nice to have but @Asleeepp is already adding plenty :)

**Comment by sovdeeth** [comment]
> > Should also cover the potential spawns api (perhaps with full entity snapshot support?).
> 
> not sure if this should be a requirement for merge. it's nice to have but @Asleeepp is already adding plenty :)

good point, though as mentioned earlier we should do the whole things in multiple prs so we can add things bit by bit

**Comment by Moderocky** [comment]
If you'd like I can create a branch in the repo that you can make multiple PRs to (and so test them together & see changes from one in another)

File: src/main/java/ch/njol/skript/effects/EffSpawnerResetTimer.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+})
+@Since("INSERT VERSION")
+@RequiredPlugins("Paper")
+public class EffSpawnerResetTimer extends Effect {
+
+	private static final boolean HAS_RESET_TIMER = Skript.methodExists(CreatureSpawner.class, "resetTimer");
+
+	static {
+		if (HAS_RESET_TIMER)
+			Skript.registerEffect(EffSpawnerResetTimer.class, "reset spawner [spawn delay] timer [of %-blocks%]");
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/effects/EffSpawnerResetTimer.java:30]: Why is blocks nullable?

File: src/main/java/ch/njol/skript/events/EvtSpawner.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+	}
+
+
+	private Literal<EntityData<?>> entityTypes;
+	private boolean isPre;
+
+	@Override
+	public boolean init(Literal<?>[] args, int matchedPattern, ParseResult parseResult) {
+		this.entityTypes =  (Literal<EntityData<?>>) args[0];
+		isPre = parseResult.hasTag("pre");
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/events/EvtSpawner.java:47]: should error if not supported

**Review by sovdeeth — CHANGES_REQUESTED** [review]
also make sure to end all test files with a new line

File: src/main/java/ch/njol/skript/expressions/ExprSpawnDelay.java [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+import org.bukkit.block.Block;
+import org.bukkit.block.CreatureSpawner;
+import org.bukkit.event.Event;
+import org.jetbrains.annotations.Nullable;
+
+@Name("Spawner Spawn Delay")
+@Description({
+	"Gets or sets the spawn delay of a Spawner block. Can also be used to get the minimum or maximum value.",
+	"This value is used when the spawner resets its delay. Measured in ticks.",
+	"When setting minimum/maximum values, the spawner will choose a random number between the minimum and maximum delay to use as it's delay."
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/expressions/ExprSpawnDelay.java:23]:
```suggestion
	"Gets or sets the spawn delay of a spawner block. Can also be used to get the minimum or maximum delays.",
	"Each reset, the spawner chooses a new delay between its min and max delays to use as the current delay."
```

File: src/main/java/ch/njol/skript/expressions/ExprSpawnDelay.java [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+})
+@Examples({
+	"set {_count} to minimum spawn delay of target block",
+	"set spawn delay of {_spawner} to 800"
+})
+@Since("INSERT VERSION")
+public class ExprSpawnDelay extends SimplePropertyExpression<Block, Integer> {
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/expressions/ExprSpawnDelay.java:30]: Needs to work in terms of timespans, not integers.

File: src/main/java/ch/njol/skript/expressions/ExprRequiredPlayerRange.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+@Name("Spawner Required Player Range")
+@Description({
+	"Gets or sets the required player range of a Spawner block.",
+	"This determines the maximum distance (squared) a player can be in order for this spawner to be active."
+})
+@Examples({
+	"set {_count} to required player range of target block",
+	"set required player range of {_spawner} to 10"
+})
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/expressions/ExprRequiredPlayerRange.java:22]: Since?

File: src/main/java/ch/njol/skript/expressions/ExprSpawnDelay.java [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+
+	static {
+		register(ExprSpawnDelay.class, Integer.class, "[:max[imum]|:min[imum]] spawn delay", "blocks");
+	}
+
+	boolean isMax;
+	boolean isMin;
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/expressions/ExprSpawnDelay.java:37]: This would probably benefit from a private Enum 
It would make logic simpler, as you could use switches.
Or add a Consumer and String value to the enum and just use those directly.

File: src/main/java/ch/njol/skript/expressions/ExprSpawnRange.java [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+import ch.njol.util.coll.CollectionUtils;
+import org.bukkit.block.Block;
+import org.bukkit.block.CreatureSpawner;
+import org.bukkit.event.Event;
+import org.jetbrains.annotations.Nullable;
+
+@Name("Spawner Spawn Range")
+@Description({
+	"Gets or sets the spawn range of a Spawner block.",
+	"This determines the radius around which the spawner will attempt to spawn mobs in."
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/expressions/ExprSpawnRange.java:18]:
```suggestion
	"Gets or sets the spawn range of a spawner block.",
	"This determines the radius around which the spawner will attempt to spawn mobs in."
```

File: src/main/java/ch/njol/skript/expressions/ExprSpawnRange.java [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+})
+@Examples({
+	"set {_count} to spawn range of target block",
+	"set spawn range of {_spawner} to 10"
+})
+@Since("INSERT VERSION")
+public class ExprSpawnRange extends SimplePropertyExpression<Block, Integer> {
+
+	static {
+		register(ExprSpawnRange.class, Integer.class, "spawn range", "blocks");
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/expressions/ExprSpawnRange.java:28]:
```suggestion
		register(ExprSpawnRange.class, Integer.class, "spawn (range|radius)", "blocks");
```

File: src/test/skript/tests/syntaxes/expressions/ExprMaxNearbyEntities.sk [hunk 9]
```diff
@@ -0,0 +1,19 @@
+test "Max Nearby Entities":
```
Comment (sovdeeth) [inline src/test/skript/tests/syntaxes/expressions/ExprMaxNearbyEntities.sk:1]: needs to test edge cases of overflow, negatives, invalid values, etc

File: src/test/skript/tests/syntaxes/expressions/ExprRequiredPlayerRange.sk [hunk 10]
```diff
@@ -0,0 +1,19 @@
+test "Required Player Range":
```
Comment (sovdeeth) [inline src/test/skript/tests/syntaxes/expressions/ExprRequiredPlayerRange.sk:1]: needs to test edge cases of overflow, negatives, invalid values, etc

File: src/test/skript/tests/syntaxes/expressions/ExprSpawnCount.sk [hunk 11]
```diff
@@ -0,0 +1,19 @@
+test "Spawn Count":
```
Comment (sovdeeth) [inline src/test/skript/tests/syntaxes/expressions/ExprSpawnCount.sk:1]: needs to test edge cases of overflow, negatives, invalid values, etc

File: src/test/skript/tests/syntaxes/expressions/ExprSpawnDelay.sk [hunk 12]
```diff
@@ -0,0 +1,43 @@
+test "Spawn Delay":
```
Comment (sovdeeth) [inline src/test/skript/tests/syntaxes/expressions/ExprSpawnDelay.sk:1]: needs to test edge cases of overflow, negatives, invalid values, etc

File: src/test/skript/tests/syntaxes/expressions/ExprSpawnRange.sk [hunk 13]
```diff
@@ -0,0 +1,19 @@
+test "Spawn Range":
```
Comment (sovdeeth) [inline src/test/skript/tests/syntaxes/expressions/ExprSpawnRange.sk:1]: needs to test edge cases of overflow, negatives, invalid values, etc

**Review by sovdeeth — CHANGES_REQUESTED** [review]
all your test messages should also omit the expected values, those are automatically added to the message.

**Comment by Asleeepp** [comment]
@burbulinis will be taking over this pr

File: src/main/java/ch/njol/skript/effects/EffSpawnerResetTimer.java [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+import ch.njol.skript.lang.Effect;
+import ch.njol.skript.lang.Expression;
+import ch.njol.skript.lang.SkriptParser.ParseResult;
+import ch.njol.util.Kleenean;
+import org.bukkit.block.Block;
+import org.bukkit.block.CreatureSpawner;
+import org.bukkit.event.Event;
+import org.jetbrains.annotations.Nullable;
+
+@Name("Spawner Reset Timer")
```
Comment (Efnilite) [inline src/main/java/ch/njol/skript/effects/EffSpawnerResetTimer.java:14]: why isnt this part of ExprSpawnDelay?

File: src/main/java/ch/njol/skript/events/EvtSpawner.java [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+
+	static {
+		List<Class<? extends Event>> events = Lists.newArrayList(SpawnerSpawnEvent.class);
+		if (HAS_PRE_SPAWNER_EVENT) {
+			events.add(PreSpawnerSpawnEvent.class);
+		}
+		Skript.registerEvent("Spawner Spawn", EvtSpawner.class, events.toArray(new Class[0]),"[:pre] spawner spawn[ing] [of %-entitydatas%]")
+			.description(
+				"Called when a spawner spawns an entity.",
+				"May also be called before a spawner spawns an entity when specified. This option requires a compatible Paper version.")
```
[the remark this answers is not in the document]
Comment (Efnilite) [reply src/main/java/ch/njol/skript/events/EvtSpawner.java:29]: unresolved

File: src/main/java/ch/njol/skript/events/EvtSpawner.java [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+	private Literal<EntityData<?>> entityTypes;
+	private boolean isPre;
+
+	@Override
+	public boolean init(Literal<?>[] args, int matchedPattern, ParseResult parseResult) {
+		this.entityTypes =  (Literal<EntityData<?>>) args[0];
+		isPre = parseResult.hasTag("pre");
+		if (isPre && !HAS_PRE_SPAWNER_EVENT) {
+			Skript.error("This event does not exist on this server version!");
+			return false;
```
Comment (Efnilite) [inline src/main/java/ch/njol/skript/events/EvtSpawner.java:50]:
```suggestion
			return false;
```

File: src/main/java/ch/njol/skript/events/EvtSpawner.java [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+		}
+		return true;
+	}
+
+	@Override
+	public boolean check(Event event) {
+		if (HAS_PRE_SPAWNER_EVENT && event instanceof PreSpawnerSpawnEvent && isPre) {
+			PreSpawnerSpawnEvent spawnerEvent = (PreSpawnerSpawnEvent) event;
```
[the remark this answers is not in the document]
Comment (Efnilite) [reply src/main/java/ch/njol/skript/events/EvtSpawner.java:58]: unresolved

File: src/main/java/ch/njol/skript/events/EvtSpawner.java [hunk 18]
```diff
@@ -0,0 +1,83 @@
...
+			for (EntityData<?> entityType : entityTypes.getArray()) {
+				if (EntityUtils.toSkriptEntityData(spawnerEvent.getType()).equals(entityType)) {
+					return true;
+				}
+			}
+			return false;
+		} else if (event instanceof SpawnerSpawnEvent && !isPre) {
+			SpawnerSpawnEvent spawnerEvent = (SpawnerSpawnEvent) event;
```
[the remark this answers is not in the document]
Comment (Efnilite) [reply src/main/java/ch/njol/skript/events/EvtSpawner.java:66]: unresolved

File: src/main/java/ch/njol/skript/events/EvtSpawner.java [hunk 19]
```diff
@@ -0,0 +1,83 @@
...
+				if (EntityUtils.toSkriptEntityData(spawnerEvent.getEntityType()).equals(entityType)) {
+					return true;
+				}
+			}
+			return false;
+		}
+		return false;
+	}
+
+
```
[the remark this answers is not in the document]
Comment (Efnilite) [reply src/main/java/ch/njol/skript/events/EvtSpawner.java:77]: unresolved

File: src/main/java/ch/njol/skript/events/EvtSpawner.java [hunk 20]
```diff
@@ -0,0 +1,83 @@
...
+			}
+			return false;
+		} else if (event instanceof SpawnerSpawnEvent && !isPre) {
+			SpawnerSpawnEvent spawnerEvent = (SpawnerSpawnEvent) event;
+			for (EntityData<?> entityType : entityTypes.getArray()) {
+				if (EntityUtils.toSkriptEntityData(spawnerEvent.getEntityType()).equals(entityType)) {
+					return true;
+				}
+			}
+			return false;
```
[the remark this answers is not in the document]
Comment (Efnilite) [reply src/main/java/ch/njol/skript/events/EvtSpawner.java:72]: unresolved

File: src/main/java/ch/njol/skript/events/EvtSpawner.java [hunk 21]
```diff
@@ -0,0 +1,83 @@
...
+		}
+		return false;
+	}
+
+
+	@Override
+	public String toString(@Nullable Event event, boolean debug) {
+		return (isPre ? "pre spawner spawn" : "spawner spawn") + (entityTypes != null ? " of " + entityTypes.toString(event, debug) : "");
```
Comment (Efnilite) [inline src/main/java/ch/njol/skript/events/EvtSpawner.java:80]: SyntaxStringBuilder

File: src/main/java/ch/njol/skript/expressions/ExprMaxNearbyEntities.java [hunk 22]
```diff
@@ -0,0 +1,84 @@
...
+import ch.njol.skript.expressions.base.SimplePropertyExpression;
+import ch.njol.util.coll.CollectionUtils;
+import org.bukkit.block.Block;
+import org.bukkit.block.CreatureSpawner;
+import org.bukkit.event.Event;
+import org.jetbrains.annotations.Nullable;
+
+@Name("Spawner Max Nearby Entities")
+@Description({
+	"Gets or sets the maximum nearby entities of a Spawner block.",
```
Comment (Efnilite) [inline src/main/java/ch/njol/skript/expressions/ExprMaxNearbyEntities.java:16]:
```suggestion
	"Gets or sets the maximum amount of entities near a spawner block.",
```

File: src/main/java/ch/njol/skript/expressions/ExprMaxNearbyEntities.java [hunk 23]
```diff
@@ -0,0 +1,84 @@
...
+	"This determines the maximum amount of similar entities that are allowed to be within spawning range of the spawner.",
+	"If there are more entities then the maximum number, the spawner will not spawn mobs."
```
Comment (Efnilite) [inline src/main/java/ch/njol/skript/expressions/ExprMaxNearbyEntities.java:18]:
```suggestion
	"If there are more entities then the maximum number, the spawner will not spawn any other entities."
```

File: src/main/java/ch/njol/skript/expressions/ExprMaxNearbyEntities.java [hunk 24]
```diff
@@ -0,0 +1,84 @@
...
+})
+@Examples({
+	"set {_count} to maximum nearby entities of target block",
+	"set max nearby entities of {_spawner} to 10"
+})
```
[the remark this answers is not in the document]
Comment (Efnilite) [reply src/main/java/ch/njol/skript/expressions/ExprMaxNearbyEntities.java:23]: unresolved

File: src/main/java/ch/njol/skript/expressions/ExprMaxNearbyEntities.java [hunk 25]
```diff
@@ -0,0 +1,84 @@
...
+	}
+
+	@Override
+	public void change(Event event, Object @Nullable [] delta, Changer.ChangeMode mode) {
+		int count = delta != null ? ((Number) delta[0]).intValue() : 0;
+
+		for (Block block : getExpr().getArray(event)) {
+			if (block.getState() instanceof CreatureSpawner spawner) {
+				int newCount = 0;
+				switch (mode) {
```
Comment (Efnilite) [inline src/main/java/ch/njol/skript/expressions/ExprMaxNearbyEntities.java:53]: enhanced switch

File: src/main/java/ch/njol/skript/expressions/ExprMaxNearbyEntities.java [hunk 26]
```diff
@@ -0,0 +1,84 @@
...
+
+	@Override
+	public Class<? extends Integer> getReturnType() {
+		return Integer.class;
+	}
+
+	@Override
+	protected String getPropertyName() {
+		return "maximum nearby entities";
+	}
```
[the remark this answers is not in the document]
Comment (Efnilite) [reply src/main/java/ch/njol/skript/expressions/ExprMaxNearbyEntities.java:83]: unresolved

File: src/main/java/ch/njol/skript/expressions/ExprRequiredPlayerRange.java [hunk 27]
```diff
@@ -0,0 +1,83 @@
...
+import ch.njol.util.coll.CollectionUtils;
+import org.bukkit.block.Block;
+import org.bukkit.block.CreatureSpawner;
+import org.bukkit.event.Event;
+import org.jetbrains.annotations.Nullable;
+
+@Name("Spawner Required Player Range")
+@Description({
+	"Gets or sets the required player range of a Spawner block.",
+	"This determines the maximum distance (squared) a player can be in order for this spawner to be active."
```
[the remark this answers is not in the document]
Comment (Efnilite) [reply src/main/java/ch/njol/skript/expressions/ExprRequiredPlayerRange.java:17]: unresolved

File: src/main/java/ch/njol/skript/expressions/ExprRequiredPlayerRange.java (hunk 5 above)
Comment (Efnilite) [reply src/main/java/ch/njol/skript/expressions/ExprRequiredPlayerRange.java:22]: unresolved

File: src/main/java/ch/njol/skript/expressions/ExprRequiredPlayerRange.java [hunk 28]
```diff
@@ -0,0 +1,83 @@
...
+})
+@Examples({
+	"set {_count} to required player range of target block",
+	"set required player range of {_spawner} to 10"
+})
+public class ExprRequiredPlayerRange extends SimplePropertyExpression<Block, Integer> {
+
+	static {
+		register(ExprRequiredPlayerRange.class, Integer.class, "required player range", "blocks");
```
[the remark this answers is not in the document]
Comment (Efnilite) [reply src/main/java/ch/njol/skript/expressions/ExprRequiredPlayerRange.java:26]: unresolved

File: src/main/java/ch/njol/skript/expressions/ExprRequiredPlayerRange.java [hunk 29]
```diff
@@ -0,0 +1,83 @@
...
+	}
+
+	@Override
+	public void change(Event event, Object @Nullable [] delta, Changer.ChangeMode mode) {
+		int count = delta != null ? ((Number) delta[0]).intValue() : 0;
+
+		for (Block block : getExpr().getArray(event)) {
+			if (block.getState() instanceof CreatureSpawner spawner) {
+				int newCount = 0;
+				switch (mode) {
```
Comment (Efnilite) [inline src/main/java/ch/njol/skript/expressions/ExprRequiredPlayerRange.java:52]: enhanced switch

File: src/main/java/ch/njol/skript/expressions/ExprRequiredPlayerRange.java [hunk 30]
```diff
@@ -0,0 +1,83 @@
...
+
+	@Override
+	public Class<? extends Integer> getReturnType() {
+		return Integer.class;
+	}
+
+	@Override
+	protected String getPropertyName() {
+		return "required player range";
+	}
```
[the remark this answers is not in the document]
Comment (Efnilite) [reply src/main/java/ch/njol/skript/expressions/ExprRequiredPlayerRange.java:82]: unresolved

File: src/main/java/ch/njol/skript/expressions/ExprSpawnCount.java [hunk 31]
```diff
@@ -0,0 +1,86 @@
...
+import ch.njol.util.coll.CollectionUtils;
+import org.bukkit.block.Block;
+import org.bukkit.block.CreatureSpawner;
+import org.bukkit.event.Event;
+import org.jetbrains.annotations.Nullable;
+
+@Name("Spawner Spawn Count")
+@Description({
+	"Gets or sets the spawn count of a Spawner block.",
+	"This determines how many mobs attempt to spawn."
```
[the remark this answers is not in the document]
Comment (Efnilite) [reply src/main/java/ch/njol/skript/expressions/ExprSpawnCount.java:18]: unresolved

File: src/main/java/ch/njol/skript/expressions/ExprSpawnCount.java [hunk 32]
```diff
@@ -0,0 +1,86 @@
...
+})
+@Examples({
+	"set {_count} to spawn count of target block",
+	"set spawn count of {_spawner} to 10"
+})
+@Since("INSERT VERSION")
+public class ExprSpawnCount extends SimplePropertyExpression<Block, Integer> {
+
+	static {
+		register(ExprSpawnCount.class, Integer.class, "spawn count", "blocks");
```
[the remark this answers is not in the document]
Comment (Efnilite) [reply src/main/java/ch/njol/skript/expressions/ExprSpawnCount.java:28]: unresolved

File: src/main/java/ch/njol/skript/expressions/ExprSpawnDelay.java (hunk 3 above)
Comment (Efnilite) [reply src/main/java/ch/njol/skript/expressions/ExprSpawnDelay.java:23]: unresolved

File: src/main/java/ch/njol/skript/expressions/ExprSpawnDelay.java (hunk 4 above)
Comment (Efnilite) [reply src/main/java/ch/njol/skript/expressions/ExprSpawnDelay.java:30]: unresolved

File: src/main/java/ch/njol/skript/expressions/ExprSpawnDelay.java (hunk 6 above)
Comment (Efnilite) [reply src/main/java/ch/njol/skript/expressions/ExprSpawnDelay.java:37]: unresolved

File: src/main/java/ch/njol/skript/expressions/ExprSpawnDelay.java [hunk 33]
```diff
@@ -0,0 +1,138 @@
...
+	@Override
+	public void change(Event event, @Nullable Object[] delta, Changer.ChangeMode mode) {
+		int count = delta != null ? ((Number) delta[0]).intValue() : 0;
+
+		for (Block block : getExpr().getArray(event)) {
+			CreatureSpawner spawner = (CreatureSpawner) block.getState();
+			if (spawner != null) {
+				int newCount = 0;
+				switch (mode) {
+					case SET:
```
Comment (Efnilite) [inline src/main/java/ch/njol/skript/expressions/ExprSpawnDelay.java:72]: enhanced switch

File: src/main/java/ch/njol/skript/expressions/ExprSpawnRange.java (hunk 7 above)
Comment (Efnilite) [reply src/main/java/ch/njol/skript/expressions/ExprSpawnRange.java:18]: unresolved

File: src/main/java/ch/njol/skript/expressions/ExprSpawnRange.java (hunk 8 above)
Comment (Efnilite) [reply src/main/java/ch/njol/skript/expressions/ExprSpawnRange.java:28]: unresolved

Comment (Efnilite) [reply src/main/java/ch/njol/skript/expressions/ExprSpawnRange.java:28]: unresolved

File: src/main/java/ch/njol/skript/expressions/ExprSpawnRange.java [hunk 34]
```diff
@@ -0,0 +1,86 @@
...
+
+	@Override
+	public void change(Event event, @Nullable Object[] delta, Changer.ChangeMode mode) {
+		int count = delta != null ? ((Number) delta[0]).intValue() : 0;
+
+		for (Block block : getExpr().getArray(event)) {
+			if (block.getState() instanceof CreatureSpawner spawner) {
+				int newCount = 0;
+				switch (mode) {
+					case REMOVE:
```
Comment (Efnilite) [inline src/main/java/ch/njol/skript/expressions/ExprSpawnRange.java:55]: enhanced switch

**Review by Efnilite — CHANGES_REQUESTED** [review]

**Comment by Asleeepp** [comment]
closed in favour of https://github.com/SkriptLang/Skript/pull/7477
